Validity Rollups on Bitcoin

john-light opened the pull request:
# ⚠️ FIRST DRAFT - POSTED FOR REVIEW ⚠️

The report will be finalized soon. Until then, please:
- Be aware that some information may be incorrect
- Leave a comment if you have any factual / typo corrections or vitally important information to add
  - Preferably, leave comments or make suggested changed inline on the markdown file

john-light committed:
[DRAFT] Validity Rollups on Bitcoin

# ⚠️ FIRST DRAFT - POSTED FOR REVIEW ⚠️

The report will be finalized soon. Until then, please:
- Be aware that some information may be incorrect
- Leave a comment if you have any factual / typo corrections or vitally important information to add
  - Preferably, leave comments or make suggested changed inline on the markdown file

john-light committed:
Fix image description formatting

john-light committed:
Fix notice

john-light committed:
Fix bit vector weight calculation

john-light committed:
Fix typos

john-light committed:
Various updates

- Rearranged some sections  
- Added new information  
- Added new citations

john-light committed:
Add section about recursive covenant transfer restrictions

john-light committed:
Lots of updates

- Updated the wording in some sections
- Added more references
- Added preface
- Added table of contents

john-light committed:
Another update

- Add network optimization section
- Add historical details about recursive covenants
- Some wording changes and typo fixes

john-light committed:
Added Appendix F

- RE: Mitigating harm from compromised cryptographic proof protocols and toxic waste

john-light committed:
Another batch of updates

- Moved some content from Appendix F to Appendix B
- Added chialisp reference

ercw commented on validity_rollups_on_bitcoin.md:
```diff
@@ -26,11 +62,11 @@ The first cryptocurrency to deploy zk proof technology to mainnet was Firo (fka
+The fundamental problem scaling researchers needed to solve was that in order to remain free from the dangers of centralization (censorship, double-spending, corruption, etc) it had to be easy for almost any cryptocurrency user to run a "full node" on the network and contribute to ensuring that the rules of the cryptocurrency consensus protocol are followed by block producers, and, if desired or  ...
 
-To keep it easy for almost any user to run a full node, the computational effort of verifying the blockchain had to be limited. This put decentralization at odds with scale. As the popularity of cryptocurrencies, especially bitcoin, grew over the years, this tension became increasingly important to resolve. While a wide variety of approaches were proposed and implemented over the years, cryptocur ...
+Onchain optimization techniques reduce the amount of resources that are required to process and store transactions that must be executed by full nodes. This can be achieved by decreasing the size in bytes or gas that transactions take up inside each block, and/or by decreasing the computational resources required to verify transactions, for example by optimizing the library used for signature ver ...
```
The comparison between sig verification libs and gas optimization feels forced -- gas optimization is user level, sig verification is core devs level.

john-light replied:
This section doesn't compare these techniques, it just puts them into the same category, because although these optimizations happen at different layers of the stack, the end result is still the same: to "reduce the amount of resources that are required to process and store transactions that must be executed by full nodes".

ercw commented on validity_rollups_on_bitcoin.md:
```diff
@@ -26,11 +62,11 @@ The first cryptocurrency to deploy zk proof technology to mainnet was Firo (fka
-Onchain optimization techniques reduce the amount of resources that are required to process and store transactions that must be executed by full nodes. This can be achieved by decreasing the size in bytes or gas that transactions take up inside each block, and/or by decreasing the computational resources required to verify transactions, for example by optimizing the library used for signature ver ...
+Network optimization scaling techniques have two main goals: one is to reduce block propagation latency, and the other is to reduce the bandwidth cost of participating as a full node on the bitcoin network.[14] Reducing block propagation latency has the benefit of improving the certainty of confirmations by reducing the number of orphaned blocks, as well as making mining more "fair" by removing t ...
 
 Sharding is a technique where a cryptocurrency's block processing and storage is split into two or more groups of nodes. Security is shared between shards so that no shard is easier to attack than any other shard. The effect of sharding is that full nodes can be confident that all of the cryptocurrency's consensus rules are being followed while only needing to store and execute a fraction of all  ...
```
> Security is shared between shards so that no shard is easier to attack than any other shard.

Seems categorically wrong, each shard has their own security and it is not shared.  It is parameterized the same, but the shard-specific circumstances are unique. One shard can be entirely Sybil'd, for instance.

john-light replied:
I'm trying to summarize the security properties described here:

> Sharding is a technique that gets you all three. A sharded blockchain is...
> 
> Secure: an attacker can't target a small part of the system with a small amount of resources; they can only try to dominate and attack the whole thing...
> 
> The random sampling prevents the attacker from concentrating their power on one shard. In a 100-chain multichain ecosystem, the attacker only needs ~0.5% of the total stake to wreak havoc: they can focus on 51% attacking a single chain. In a sharded blockchain, the attacker must have close to ~30-40% of the entire stake to do the same (in other words, the chain has shared security). Certainly, they can wait until they get lucky and get 51% in a single shard by random chance despite having less than 50% of the total stake, but this gets exponentially harder for attackers that have much less than 51%. If an attacker has less than ~30%, it's virtually impossible.

https://vitalik.ca/general/2021/04/07/sharding.html

However later in that post, Vitalik also says:

> Sharded blockchains depend on sharded peer-to-peer networks, and each individual p2p "subnet" is easier to attack because it has fewer nodes. The [subnet model used for data availability sampling](https://hackmd.io/@vbuterin/sharding_proposal#Blob-publication-process) mitigates this because there is some redundancy between subnets, but even still there is a risk.

Is that what you're describing when you say "One shard can be entirely Sybil'd"?

delbonis replied:
It's not possible to choose the shard one validates on so as long as there's a sufficient total number of validator accounts it's statically unlikely that any malicious subset with less than 1/3 of the total stake will end up with >1/3 on any one shard.  And all that would accomplish is preventing finality, it doesn't allow reversal of finalized shard blocks, which requires 2/3.  This logic I think only applies to data availability sharding that they're doing and not execution sharding.

john-light replied:
@delbonis  thanks for that. So would you say my initial characterization was accurate? Or I should remove this sentence since it doesn't apply to all kinds of sharding?

> Security is shared between shards so that no shard is easier to attack than any other shard.

john-light committed:
Add Appendix F name

john-light committed:
A couple more updates

- Added information about recent regulatory crackdowns on privacy protocols
- Added conclusion section

john-light committed:
Add PoB reference

john-light committed:
References fix #1

delbonis commented on validity_rollups_on_bitcoin.md:
```diff
@@ -292,113 +337,161 @@ Although Lightning is most famous for settling on the bitcoin mainchain, Lightni
+Less obvious is the need for recursive covenants, at least in the Del Bonis bitcoin rollup design. Recursive covenants are a type of smart contract that restricts the type of script that satoshis can be sent to once they are spent. Del Bonis uses recursive covenenants to ensure that satoshis that are locked in a rollup script and haven't been withdrawn by their owner yet remain in the script from ...
 
-Recursive covenants are a change to Script that has long been considered by the bitcoin community. As of February 2022, Dave Harding reports that, "I believe the last time the merits of allowing recursive covenants was discussed at length on this list, not a single person replied to say that they were opposed to the idea."[90] However there are currently no specific proposals that have achieved b ...
+Recursive covenants are a change to Script that has long been considered by the bitcoin community.[99, 100, 101] However there are currently no specific proposals that have achieved broad consensus among the bitcoin developer community to implement recursive covenants. There are proposals such as BIP-118 and BIP-119 that enable more limited covenants, but these do not have the recursion property  ...
```
I think of it more as the recursion is more for being able to propagate the construction forwards, and being able to consume user funds and add them into the state tree is part of that.

john-light replied:
Thanks, updated the wording here.

delbonis commented on validity_rollups_on_bitcoin.md:
```diff
@@ -292,113 +337,161 @@ Although Lightning is most famous for settling on the bitcoin mainchain, Lightni
+Recursive covenants are a change to Script that has long been considered by the bitcoin community.[99, 100, 101] However there are currently no specific proposals that have achieved broad consensus among the bitcoin developer community to implement recursive covenants. There are proposals such as BIP-118 and BIP-119 that enable more limited covenants, but these do not have the recursion property  ...
 
-The OP_EVAL and PUSHSCRIPT opcodes are nice-to-haves that reduce the size of the rollup script in some areas, reducing the amount of blockspace used and therefore making rollups cheaper to use, all else being equal. Increasing or removing the stack element size limit is another nice-to-have that makes rollups cheaper to use, in this case by increasing the number of transactions that can fit into  ...
+Del Bonis identifies the OP_EVAL and PUSHSCRIPT opcodes as nice-to-haves that reduce the size of the rollup script in some areas, reducing the amount of blockspace used and therefore making rollups cheaper to use, all else being equal. Increasing or removing the stack element size limit are another nice-to-have that he identifies as making rollups cheaper to use, in this case by increasing the nu ...
```
The stack element size limit is probably the most important single change since it would just be so hard and error-prone to try to work around it.  The other limitations we can deal with but being unable to build moderately-large public inputs to  pass to a zkp verification opcode without wading through a ton of stack elements would probably be a killer.

This might be enough of a change though to force us to wait for Simplicity or more momentum behind Chialisp instead of trying to make it work in the Bitcoin script interpreter since it seems like a stretch.

john-light replied:
Thanks, updated the wording here to emphasize more the importance of increasing the stack element size limit.

delbonis commented on validity_rollups_on_bitcoin.md:
```diff
@@ -292,113 +337,161 @@ Although Lightning is most famous for settling on the bitcoin mainchain, Lightni
 To answer the question of whether validity rollups on bitcoin create new opportunities for MEV, we must first be specific about what kind of validity rollups we are enabling on bitcoin. It is possible to limit the expressivity of the scripting capabilities supported by validity rollups that can be built on bitcoin by limiting the complexity of the proofs that bitcoin full nodes are able to verify ...
 
-The bitcoin community could also decide to enable more expressive validity rollups to be built on bitcoin. Perhaps these rollups would be expressive enough to enable the types of contracts that are vulnerable to MEV. In this case, there would be new MEV opportunities created on bitcoin. This MEV would mainly be captured by rollup block producers on Layer 2. Due to bitcoin's long block time, it wo ...
+The bitcoin community could also decide to enable more expressive validity rollups to be built on bitcoin. Perhaps these rollups would be expressive enough to enable the types of contracts that are vulnerable to MEV. In this case, there would be new MEV opportunities created on bitcoin. This MEV would mainly be captured by rollup block producers on L2. Due to bitcoin's long block time, it would b ...
```
>In remains

(typo)

john-light replied:
Good eye! Fixed

delbonis commented on validity_rollups_on_bitcoin.md:
```diff
@@ -292,113 +337,161 @@ Although Lightning is most famous for settling on the bitcoin mainchain, Lightni
+
+Some commentators have cited the recursive restrictions enabled by recursive covenants as justification for pushing back on the implementation of covenants in bitcoin.[111, 112] They point out that this capability could be used by totalitarian governments to force bitcoin users in their jurisdictions to lock their satoshis into recursive covenants where the government has programmed in their cont ...
+
+It is currently possible for governments to implement restrictions on how the satoshis owned by their citizens can be transferred, and in a much more flexible way than recursive covenants allow, using a type of smart contract that has been standard in bitcoin for over a decade: multisig.[113] The basic idea is that the government would require satoshis owned by their citizens to be encumbered by  ...
```
>government copmuter

(typo, or is it?)

john-light replied:
freudian slip 😆

delbonis commented on validity_rollups_on_bitcoin.md:
```diff
@@ -518,9 +611,13 @@ Zcash is the first implementation of the Zerocash protocol. The defining privacy
+
+Some people may argue that the "opt-in" nature of encrypted protocols implemented in a validity rollup makes the privacy less effective for those who use it, and that "opt-out" encrypted protocols on blockchains such as Beam or Monero will always provide superior privacy, all else being equal. This is a misunderstanding. There is actually no such thing as purely "opt-out" privacy, because _all_ p ...
 
 ## Appendix C. Increasing throughput with offchain data availability
```
In the third paragraph of Appendix C:

>at the same time as support for validity rollups is added by implementing flexible enough to support various offchain data availability solutions

More of a readability comment, my brain is having a hard time parsing this.

john-light replied:
Oh yeah, good catch. There was definitely a word or two missing there. Updated to say:

> Bitcoin could add support for these validia protocols at the same time that support for validity rollups is added by implementing flexible enough rollup scripts to support various offchain data availability solutions.

delbonis commented on validity_rollups_on_bitcoin.md:
```diff
@@ -647,6 +746,20 @@ Proofs of the validity of the blockchain solve the IBD issue for blocks of arbit
 
 This "proof-sync" technique may solve the IBD problem of large blocks, but it does not solve the mining centralization problems. Work still needs to be done to decrease the bandwidth costs and propogation times of increased Layer 1 transaction volume, for example by implementing or further optimizing protocols such as Compact Block Relay and Erlay that reduce bandwidth requirements even as transa ...
```
This would be really cool to expand on if you think it fits.  I'm assuming they build some kind of state commitment?  Is this generic enough that we could eventually soft fork in the state roots?  Although since it's not actually a consensus change I suppose you could gossip the state roots and do it optimistically/incrementally like utreexo sync.

john-light replied:
I wish I could! But I don't know exactly how they work, and I think it's too early and too tangential for this report to spend any more time on the topic than I have already. Given that there are two "STARK-verified bitcoin" projects that I'm aware of (Zerosync and Khepri) it's [only a matter of time until some long-form documentation comes out](https://twitter.com/dimahledba/status/1565382982382272513) about them. I appreciate the encouragement but won't be expanding on this concept any further in this report :)

delbonis commented on validity_rollups_on_bitcoin.md:
```diff
@@ -280,7 +325,7 @@ The onchain transactions needed to open and settle (and occasionally rebalance)
 The ability to layer validity rollups on top of one another also means that capacity for Lightning can be added as-needed. Each time a rollup reaches its maximum capacity, another rollup could be deployed as an additional layer, then Lightning onboarding can continue until that layer fills up, and so on.
 
-Although Lightning is most famous for settling on the bitcoin mainchain, Lightning channels can be opened on any Lightning-compatible blockchain. Additionally, if there is a Lightning node with channels open on two or more chains, it can route payments between users of these different chains, as if those users had channels open on the same chain. In cases where there are different assets being se ...
+Although Lightning is most famous for settling on the bitcoin mainchain, Lightning channels can be opened on any Lightning-compatible blockchain. Additionally, if there is a Lightning node with channels open on two or more chains, it can route payments between users of these different chains, as if those users had channels open on the same chain. In cases where there are different assets being se ...
```
Not sure if you want to go into it much but it could be interesting to mention that in a lightning-oriented rollup the open/close L1 data costs could be much smaller, beyond just the lack of signatures.

john-light replied:
> it could be interesting to mention that in a lightning-oriented rollup the open/close L1 data costs could be much smaller, beyond just the lack of signatures.

how so?

delbonis replied:
If we assume that "funds at rest" are stored in an account-like structure then a channel open would just involve 8 bytes for the indexes of two parties, plus you could do 2 bytes for the channel size (we can get by with a fixed list of sizes), and maybe plus ~2 more bytes for any metadata and fee information.  We can assume either even dualfunding or the first of the two parties fronts the full balance.

Storing channels could be done either with sequential indexes or some commitment to each party.  In the worst case, close txs would involve 8 bytes for the two indexes of each party again, the ~2 bytes for the meta+fee, probably 4 bytes for the end balance, and then for each HTLC 20 bytes for the hash, 4 bytes for the balance and direction, and 4 bytes for the timeout.  HTLCs can get stored in their own indexes table and remember who is allowed to claim them so that information doesn't need to be included in the claim tx.

The HTLC claims are just the 4 byte index of the HTLC, a 20 byte preimage, and probably another ~2 bytes for meta+fee.  It's possible you could even make timed-out HTLCs reclaim themselves automatically.

john-light replied:
Ah yes, I had considered including an account-based version of the Lightning rollup for completeness, but wasn't sure about the transaction size details. So Table 3 and 4 would be modified to look like this:

#### Table 3. Comparing the size of different types of UTXO model dual-funded Lightning channel open transactions

| Transaction type                       | Non-discounted data weight (4x) | Discounted data weight (1x) | Total (WU) |
| -------------------------------------- |:-------------------------------:|:---------------------------:|:----------:|
| Rollup LN channel open (account model) | 0                               | 12                          | 12         |
| Rollup LN channel open (UTXO model)    | 0                               | 197.5                       | 197.5      |
| Mainchain LN channel open              | 795                             | 211                         | 1006       |

> Table 3 description: The first rollup transaction is a dual-funded Lightning channel open transaction using an account model rollup. The second rollup transaction is a 2-P2WPKH-input-1-P2WSH-output-2-P2WPKH-output transaction with the 27 byte witnesses stripped from each input. The mainchain transaction is a 2-P2WPKH-input-1-P2WSH-output-2-P2WPKH-output transaction.[^96]

#### Table 4. Comparing the per-block capacity of different types of UTXO model dual-funded Lightning channel open transactions

| Transaction type                       | 1,000,000 WU                          | 3,000,0000 WU            | Total (tx/block) |
| -------------------------------------- |:-------------------------------------:|:------------------------:|:----------------:|
| Rollup LN channel open (account model) | Validity proof, script, other tx data | 250,000 txs (12 WU/tx)   | 250,000          |
| Rollup LN channel open (UTXO model)    | Validity proof, script, other tx data | 15,190 txs (197.5 WU/tx) | 15,190           |
| Mainchain LN channel open              | 994 txs (1006 WU/tx)                  | 2,982 txs (1006 WU/tx)   | 3,976            |

> Table 4 description: The block is broken up into 1,000,000 WU and 3,000,000 WU portions to illustrate how in the case of rollup transactions, some of the weight limit is reserved for the validity proof and script data while the rest can be used for the rollup transaction data.

Is that right @delbonis?

delbonis reviewed (commented):
🎉⚡

john-light committed:
References fix #2

- Plus some wording changes here and there

john-light committed:
Fix typo

john-light committed:
Update some wording and fix typos

- Thanks to Trey Del Bonis for the review comments that led to these changes

john-light committed:
Add Khepri

- Also fix some markdown formatting

john-light committed:
Add bidirectional reference links

- Adjust some other wording
- Fix some typos

john-light committed:
Misc. updates

- Add TOC, section, subsection, table, and figure links  
- Small wording changes throughout  
- Fix some typos

sebareca commented on validity_rollups_on_bitcoin.md:
```diff
@@ -86,138 +139,147 @@ The ability to freely transfer assets on a rollup to any other rollup address ma
+In both cases, these alternative programming languages would not be able to operate on BTC natively. Instead, a bridge (also referred to as a "two-way peg" in the literature) would need to be built that would lock BTC on the bitcoin mainchain, release an equivalent amount of BTC IOUs inside the new execution environment, and implement some mechanism for converting the IOUs back into BTC on the ma ...
 
-### 3.2 New privacy protections
+If it were possible to build validity rollups with alternative execution environments on bitcoin, then bitcoin users could experiment with these alternative execution environments while retaining the full "self-custodial" ownership security of bitcoin L1. Validity rollups would eliminate the need for BTC IOUs, since validity rollups built on bitcoin would be able to operate on BTC nativelyAdditio ...
```
There's a missing full stop after "...operate on BTC natively".

john-light replied:
Good eye! Thanks

john-light committed:
Fix typo

- Add acknowledgement of the typo finder

john-light committed:
Fix more typos

- Also some minor wording adjustments

john-light committed:
Add ideas for limiting the number of proofs per block

- Credit to Ruben Somsen for the idea to increase the weight of the validity proofs

john-light committed:
Fix typos and add some info

- Credit to Pete Eyre for finding some of these typos
- Credit to Ruben Somsen for pointing out how L1 blocks full of rollup updates would be larger than current average block sizes and that this should be pointed out in the costs section

john-light committed:
Fix previous update

- Accidentally added info that is currently under review

john-light committed:
Update internal link format

john-light committed:
Update table links

john-light committed:
Update conclusion with costs associated with higher throughput via rollups

john-light committed:
Remove draft notice
